### PR TITLE
Fix/mnt 22293 limit on declare version as record

### PR DESCRIPTION
**Pull request by brijmohan1**

### CODE REVIEW

File: repository/src/main/java/org/alfresco/repo/domain/node/AbstractNodeDAOImpl.java [hunk 1]
```diff
@@ -3780,22 +3780,35 @@ public void done()
...
+                    // Remember exception we want to throw and exit. If we throw within here, it will be wrapped by IBatis
+                    toThrow = new CyclicChildRelationshipException(
+                            "Child Association Cycle detected hitting nodes: " + nodeIds,
+                            childAssociationRef);
+                    return false;
+                }
             }
-            cycleCheck(nodeId);
-            nodeIds.remove(nodeId);
+            // cycleCheck(nodeId);
```
Comment (suneet-gupta) [inline repository/src/main/java/org/alfresco/repo/domain/node/AbstractNodeDAOImpl.java:3810]: we shd remove commented code

**Review by suneet-gupta — CHANGES_REQUESTED** [review]
Please do not touch original formatting or add/remove spaces if you are not making any changes at that place.

**Review by kavitshah-gl — APPROVED** [review]
This seems fine.

File: repository/src/main/resources/alfresco/ibatis/org.alfresco.repo.domain.dialect.Dialect/node-common-SqlMap.xml [hunk 2]
```diff
@@ -953,7 +953,47 @@
...
+               assoc.assoc_index         as assoc_index
+        from alf_child_assoc assoc
+                 join childAssosList on (assoc.parent_node_id = childAssosList.childNodeId)
+                 join alf_node parentNode on (parentNode.id = assoc.parent_node_id)
+                 join alf_store parentStore on (parentStore.id = parentNode.store_id)
+                 join alf_node childNode on (childNode.id = assoc.child_node_id)
+                 left join alf_store childStore on (childStore.id = childNode.store_id)
+            )
+        select *
+        from childAssosList
```
Comment (tpage-alfresco) [inline repository/src/main/resources/alfresco/ibatis/org.alfresco.repo.domain.dialect.Dialect/node-common-SqlMap.xml:994]: Typo - "Assos"

File: repository/src/main/resources/alfresco/ibatis/org.alfresco.repo.domain.dialect.Dialect/node-common-SqlMap.xml [hunk 3]
```diff
@@ -953,7 +953,47 @@
...
+        <include refid="alfresco.node.select_ChildAssocsOfParent_Snippet"></include>
+        <include refid="alfresco.node.select_ChildAssocsOfParentRecursiveQuery"></include>
+    </select>
+    <sql id="select_ChildAssocsOfParentWithRecursive">
+        WITH RECURSIVE childAssosList AS (
+    </sql>
+    <sql id="select_ChildAssocsOfParentRecursiveQuery">
+        UNION ALL
+
+        select assoc.id                  as id,
```
Comment (tpage-alfresco) [inline repository/src/main/resources/alfresco/ibatis/org.alfresco.repo.domain.dialect.Dialect/node-common-SqlMap.xml:968]: I'm concerned about what this returns. If we're adding a node to the root of the repository then does it return the whole repository?  What if we move a folder near the root of the repository to a new location?

**Comment by tpage-alfresco** [comment]
I think the issue in general is not something that can sensibly be solved. I think to check for cycles in the general case then we have to traverse almost every parent-child relationship in the repository.

I noticed this comment:
```
     * TODO: When is it cheaper to go up and when is it cheaper to go down?
     *       Look at using direct queries to pass through layers both up and down.
```

For the problem we're trying to solve then I think we know there's no cycle because the node has no children?  If this is the case then we could solve the problem by changing the logic to do:
```
If the child has no children
  // No cycle
else
  // Perform existing check
```

This would improve performance for the majority of cases (creating a new child) and make it slightly worse when moving a folder or adding a secondary parent association to a folder.

**Comment by jottley** [comment]
The related JIRA was closed as Won't Do.  Closing the PR

**Pull request closed without merging**